File: lib/javascript/src/subscription-metadata.ts
import {
	assertChecksum,
	assertInRange,
	assertLength,
	BLOCK_SIZE,
	readUint16,
	withChecksum,
	writeUint16,
} from "./bytes.ts";
import { decodeDate, encodeDate, type Date16Bit } from "./date.ts";

/** Block 12 or 16 of a personal card. */
export type SubscriptionMetadata = {
	/** Byte 0, tells the products of a card apart. */
	productId: number;
	/** Byte 1, always `1` so far. */
	unknown1: number;
	/** When the product was bought. */
	purchasedAt: Date16Bit;
	/** Bytes 4 to 7, always `00210000` so far. */
	unknown2: Uint8Array;
	/** How many days the product is valid for. */
	validityDays: number;
	/** Bytes 10 to 14, always `0021000000` so far. */
	unknown3: Uint8Array;
};

/**
 * Decodes a subscription metadata block.
 * @param block The 16-byte block.
 */
export function decodeSubscriptionMetadata(
	block: Uint8Array,
): SubscriptionMetadata {
	assertLength(block, BLOCK_SIZE, "subscription metadata block");
	assertChecksum(block);
	return {
		productId: block[0]!,
		unknown1: block[1]!,
		purchasedAt: decodeDate(block.subarray(2, 4)),
		unknown2: block.slice(4, 8),
		validityDays: readUint16(block, 8),
		unknown3: block.slice(10, 15),
	};
}

/** Encodes subscription metadata into a 16-byte block. */
export function encodeSubscriptionMetadata(
	metadata: SubscriptionMetadata,
): Uint8Array {
	assertInRange("product id", metadata.productId, 0, 0xff);
	assertInRange("unknown1", metadata.unknown1, 0, 0xff);
	assertInRange("validity days", metadata.validityDays, 0, 0xffff);
	assertLength(metadata.unknown2, 4, "unknown2");
	assertLength(metadata.unknown3, 5, "unknown3");
	const block = new Uint8Array(BLOCK_SIZE);
	block[0] = metadata.productId;
	block[1] = metadata.unknown1;
	block.set(encodeDate(metadata.purchasedAt), 2);
	block.set(metadata.unknown2, 4);
	writeUint16(block, 8, metadata.validityDays);
	block.set(metadata.unknown3, 10);
	return withChecksum(block);
}

File: lib/rust/src/subscription_metadata.rs
use crate::bytes::{BLOCK_SIZE, Block, check_checksum, chunk, with_checksum};
use crate::date::Date;
use crate::error::Result;

/// Block 12 or 16 of a personal card, one product per sector.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct SubscriptionMetadata {
	/// Byte 0, tells the products of a card apart.
	pub product_id: u8,
	/// Byte 1, always `1` so far.
	pub unknown1: u8,
	/// Bytes 2 and 3.
	pub purchased_at: Date,
	/// Bytes 4 to 7, always `00210000` so far.
	pub unknown2: [u8; 4],
	/// Bytes 8 and 9, big-endian.
	pub validity_days: u16,
	/// Bytes 10 to 14, always `0021000000` so far.
	pub unknown3: [u8; 5],
}

impl SubscriptionMetadata {
	/// Decodes block 12 or 16.
	///
	/// # Errors
	/// [`Error::Checksum`](crate::Error::Checksum) or the errors of [`Date::decode`].
	pub fn decode(block: &Block) -> Result<Self> {
		check_checksum(block)?;
		let [product_id, unknown1, d0, d1, ..] = *block;
		Ok(Self {
			product_id,
			unknown1,
			purchased_at: Date::decode([d0, d1])?,
			unknown2: chunk(block, 4),
			validity_days: u16::from_be_bytes(chunk(block, 8)),
			unknown3: chunk(block, 10),
		})
	}

	/// Encodes into block 12 or 16 with its checksum.
	///
	/// # Errors
	/// The errors of [`Date::encode`].
	pub fn encode(self) -> Result<Block> {
		let mut block = [0; BLOCK_SIZE];
		block[0] = self.product_id;
		block[1] = self.unknown1;
		block[2..4].copy_from_slice(&self.purchased_at.encode()?);
		block[4..8].copy_from_slice(&self.unknown2);
		block[8..10].copy_from_slice(&self.validity_days.to_be_bytes());
		block[10..15].copy_from_slice(&self.unknown3);
		Ok(with_checksum(block))
	}
}
